app/services/katello/pulp/smart_proxy_repository.rb

```diff
@@ -0,0 +1,179 @@
+module Katello
+  module Pulp
+    class SmartProxyRepository
+      attr_accessor :smart_proxy
+
+      def initialize(smart_proxy)
+        @smart_proxy = smart_proxy
+      end
+
+      def lifecycle_environments(organization_id = nil)
+        scope = @smart_proxy.lifecycle_environments
+        scope = scope.where(organization_id: organization_id) if organization_id
+        scope
+      end
+
+      def add_lifecycle_environment(environment)
+        @smart_proxy.lifecycle_environments << environment
+      end
+
+      def remove_lifecycle_environment(environment)
+        @smart_proxy.lifecycle_environments.find(environment.id)
+        unless @smart_proxy.lifecycle_environments.destroy(environment)
+          fail _("Could not remove the lifecycle environment from the smart proxy")
+        end
+      rescue ActiveRecord::RecordNotFound
+        raise _("Lifecycle environment was not attached to the smart proxy; therefore, no changes were made.")
+      end
+
+      def available_lifecycle_environments(organization_id = nil)
+        scope = Katello::KTEnvironment.not_in_capsule(@smart_proxy)
+        scope = scope.where(organization_id: organization_id) if organization_id
+        scope
+      end
```

Comment:
same comment for this method

Reply:
Really any method that doesn't depend on data from the pulp server, should be moved to the smart proxy class (will stop adding repeated comments)